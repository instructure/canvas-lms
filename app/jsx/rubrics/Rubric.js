/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import _ from 'lodash'
import PropTypes from 'prop-types'
import Table from '@instructure/ui-elements/lib/components/Table'
import Flex, { FlexItem } from '@instructure/ui-layout/lib/components/Flex'
import I18n from 'i18n!edit_rubric'

import Criterion from './Criterion'

import { rubricShape, rubricAssessmentShape, rubricAssociationShape } from './types'
import { roundIfWhole } from './Points'

const totalString = (score) => I18n.t('Total Points: %{total}', {
  total: I18n.toNumber(score, { precision: 1 })
})

const totalAssessingString = (score, possible) =>
  I18n.t('Total Points: %{total} out of %{possible}', {
    total: roundIfWhole(score),
    possible: I18n.toNumber(possible, { precision: 1 })
  })

const Rubric = (props) => {
  const {
    allowExtraCredit,
    customRatings,
    onAssessmentChange,
    rubric,
    rubricAssessment,
    rubricAssociation,
    isSummary
  } = props
  const assessing = onAssessmentChange !== null
  const priorData = _.get(rubricAssessment, 'data', [])
  const byCriteria = _.keyBy(priorData, (ra) => ra.criterion_id)
  const criteriaById = _.keyBy(rubric.criteria, (c) => c.id)
  const allComments = _.get(rubricAssociation, 'summary_data.saved_comments', {})
  const hidePoints = _.get(rubricAssociation, 'hide_points', false)
  const freeForm = rubric.free_form_criterion_comments

  const onCriteriaChange = (id) => (update) => {
    const data = priorData.map((prior) => (
      prior.criterion_id === id ? { ...prior, ...update } : prior
    ))

    const ignore = (c) => _.isUndefined(c) ? true : c.ignore_for_scoring
    const points = data
      .filter((result) => !ignore(criteriaById[result.criterion_id]))
      .map((result) => result.points !== null ? result.points : 0)

    onAssessmentChange({
      ...rubricAssessment,
      data,
      score: _.sum(points)
    })
  }

  // we show the last column for points or comments button
  const showPointsColumn = () => {
    if (isSummary) { return false }
    if (!hidePoints) { return true }
    if (assessing && !freeForm) { return true } // comments button
    return false
  }

  const criteria = rubric.criteria.map((criterion) => {
    const assessment = byCriteria[criterion.id]
    return (
      <Criterion
        allowExtraCredit={allowExtraCredit}
        key={criterion.id}
        assessment={assessment}
        criterion={criterion}
        customRatings={customRatings}
        freeForm={freeForm}
        isSummary={isSummary}
        onAssessmentChange={assessing ? onCriteriaChange(criterion.id) : undefined}
        savedComments={allComments[criterion.id]}
        hidePoints={hidePoints}
        hasPointsColumn={showPointsColumn()}
      />
    )
  })

  const possible = rubric.points_possible
  const points = _.get(rubricAssessment, 'score', possible)
  const total = assessing ? totalAssessingString(points, possible) : totalString(points)
  const hideScoreTotal = _.get(rubricAssociation, 'hide_score_total') === true
  const noScore = _.get(rubricAssociation, 'score') === null
  const showTotalPoints = !hidePoints && !hideScoreTotal

  return (
    <div className="react-rubric">
      <Table caption={rubric.title}>
        <thead>
          <tr>
            <th scope="col">{I18n.t('Criteria')}</th>
            <th scope="col">{I18n.t('Ratings')}</th>
            {
              showPointsColumn() && (
                <th scope="col">{I18n.t('Pts')}</th>
              )
            }
          </tr>
        </thead>
        <tbody className="criterions">
          {criteria}
          { showTotalPoints && (
            <tr>
              <td colSpan="3">
                <Flex justifyItems="end">
                  <FlexItem>
                    {hideScoreTotal || noScore ? null : total}
                  </FlexItem>
                </Flex>
              </td>
            </tr>
          )}
        </tbody>
      </Table>
    </div>
  )
}
Rubric.propTypes = {
  allowExtraCredit: PropTypes.bool,
  customRatings: PropTypes.arrayOf(PropTypes.object),
  onAssessmentChange: PropTypes.func,
  rubric: PropTypes.shape(rubricShape).isRequired,
  rubricAssessment: (props) => {
    const shape = PropTypes.shape(rubricAssessmentShape)
    const rubricAssessment = props.onAssessmentChange ? shape.isRequired : shape
    return PropTypes.checkPropTypes({ rubricAssessment }, props, 'prop', 'Rubric')
  },
  rubricAssociation: PropTypes.shape(rubricAssociationShape),
  isSummary: PropTypes.bool
}
Rubric.defaultProps = {
  allowExtraCredit: false,
  customRatings: [],
  onAssessmentChange: null,
  rubricAssessment: null,
  rubricAssociation: {},
  isSummary: false
}

export default Rubric
