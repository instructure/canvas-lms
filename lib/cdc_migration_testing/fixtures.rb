# frozen_string_literal: true

#
# Copyright (C) 2020 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

Dir.glob(File.join(__dir__, "fixtures/**/*.rb")).each do |file|
  require_relative file
end

Dir.glob(File.join(__dir__, "../../gems/plugins/*/lib/cdc_migration_testing/fixtures/**/*.rb")).each do |file|
  require_relative file
end
