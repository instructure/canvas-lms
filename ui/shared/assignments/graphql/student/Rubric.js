/*
 * Copyright (C) 2019 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {arrayOf, bool, number, shape, string} from 'prop-types'
import gql from 'graphql-tag'
import {RubricCriterion} from './RubricCriterion'

export const Rubric = {
  fragment: gql`
    fragment Rubric on Rubric {
      criteria {
        ...RubricCriterion
      }
      _id
      free_form_criterion_comments: freeFormCriterionComments
      hide_score_total: hideScoreTotal
      points_possible: pointsPossible
      title
    }
    ${RubricCriterion.fragment}
  `,

  shape: shape({
    id: string.isRequired,
    criteria: arrayOf(RubricCriterion.shape),
    free_form_criterion_comments: bool,
    hide_score_total: bool,
    points_possible: number.isRequired,
    title: string.isRequired
  })
}

export const DefaultMocks = {
  Rubric: () => ({
    _id: '1',
    criteria: [{}],
    freeFormCriterionComments: false
  })
}
