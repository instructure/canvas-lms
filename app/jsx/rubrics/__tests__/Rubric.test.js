/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import sinon from 'sinon'
import { shallow } from 'enzyme'
import _ from 'lodash'
import Rubric from '../Rubric'
import { rubric, pointsAssessment } from './fixtures'

describe('the Rubric component', () => {
  it('renders as expected', () => {
    const modal = shallow(
      <Rubric
        rubric={rubric}
        rubricAssessment={pointsAssessment}
        rubricAssociation={pointsAssessment.rubric_association}
      />
    )
    expect(modal.debug()).toMatchSnapshot()
  })

  it('renders properly with no assessment', () => {
    const modal = shallow(
      <Rubric
        rubric={rubric}
        rubricAssociation={pointsAssessment.rubric_association}
      />
    )
    expect(modal.debug()).toMatchSnapshot()
  })

  const setCloned = (object, path, value) => _.setWith(_.clone(object), path, value, _.clone)
  it('hides the score total when needed', () => {
    const hidden = setCloned(pointsAssessment, 'rubric_association.hide_score_total', true)
    const modal = shallow(
      <Rubric
        rubric={rubric}
        rubricAssessment={hidden}
        rubricAssociation={hidden.rubric_association}
      />)
    expect(modal.debug()).toMatchSnapshot()
  })

  it('updates the total score when an individual criterion point assessment changes', () => {
    const onAssessmentChange = sinon.spy()
    const renderAssessing = (assessment) =>
      shallow(
        <Rubric
          onAssessmentChange={onAssessmentChange}
          rubric={rubric}
          rubricAssessment={assessment}
          rubricAssociation={assessment.rubric_association}
        />
      )

    const el = renderAssessing(pointsAssessment)
    const updated = { ...pointsAssessment.data[0], points: 2 }
    el.find('Criterion').first().prop('onAssessmentChange')(updated)

    expect(onAssessmentChange.args).toEqual([
      [{
        ...pointsAssessment,
        data: [updated, pointsAssessment.data[1]],
        score: 2 + pointsAssessment.data[1].points
      }]
    ])

    expect(renderAssessing(onAssessmentChange.args[0][0]).debug()).toMatchSnapshot()
  })

  describe('points column', () => {
    const hasPointsColumn = (expected, { rubricProps = {}, assessmentProps = {}, associationProps = {}, ...otherProps }) => {
      const el = shallow(
        <Rubric
          rubric={{...rubric, ...rubricProps}}
          rubricAssessment={{...pointsAssessment, ...assessmentProps}}
          rubricAssociation={{...pointsAssessment.association, ...associationProps}}
          onAssessmentChange={() => {}}
          {...otherProps}
        />
      )
      expect(el.find('th')).toHaveLength(expected ? 3 : 2)
      expect(el.find('Criterion').at(0).prop('hasPointsColumn')).toBe(expected)
    }

    it('does not have a points column in summary mode', () => {
      hasPointsColumn(false, { isSummary: true })
    })

    it('has a points column if points visible', () => {
      hasPointsColumn(true, {})
    })

    it('does not have a points column if points hidden and freeform', () => {
      hasPointsColumn(false, { rubricProps: { free_form_criterion_comments: true }, associationProps: { hide_points: true }})
    })

    it('does not have a points column if points hidden and not assessing', () => {
      hasPointsColumn(false, { associationProps: { hide_points: true }, onAssessmentChange: null })
    })

    it('does have a points column if points hidden, not freeform, and assessing', () => {
      hasPointsColumn(true, { associationProps: { hide_points: true }})
    })
  })

  it('ignores criteria scores when flagged as such', () => {
    const ignoreOutcomeScore = setCloned(rubric, 'criteria.1.ignore_for_scoring', true)
    const onAssessmentChange = sinon.spy()
    const ignored = { ...pointsAssessment.data[1], points: 2, criterion_id: '_invalid' }
    const assessment = setCloned(pointsAssessment, 'data.2', ignored)
    const el = shallow(
      <Rubric
        onAssessmentChange={onAssessmentChange}
        rubric={ignoreOutcomeScore}
        rubricAssessment={assessment}
        rubricAssociation={assessment.rubric_association}
      />
    )

    const updated = { ...assessment.data[1], points: 2 }
    el.find('Criterion').at(1).prop('onAssessmentChange')(updated)

    expect(onAssessmentChange.args).toEqual([
      [{
        ...assessment,
        data: [assessment.data[0], updated, ignored],
        score: assessment.data[0].points
      }]
    ])
  })
})
