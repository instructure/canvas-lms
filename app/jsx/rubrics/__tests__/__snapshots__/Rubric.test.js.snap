// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Rubric component hides the score total when needed 1`] = `
"<div className=\\"react-rubric\\">
  <Table caption=\\"Point Rubric\\" hover={false} size=\\"medium\\">
    <thead>
      <tr>
        <th scope=\\"col\\">
          Criteria
        </th>
        <th scope=\\"col\\">
          Ratings
        </th>
        <th scope=\\"col\\">
          Pts
        </th>
      </tr>
    </thead>
    <tbody className=\\"criterions\\">
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={{...}} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={{...}} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
    </tbody>
  </Table>
</div>"
`;

exports[`the Rubric component renders as expected 1`] = `
"<div className=\\"react-rubric\\">
  <Table caption=\\"Point Rubric\\" hover={false} size=\\"medium\\">
    <thead>
      <tr>
        <th scope=\\"col\\">
          Criteria
        </th>
        <th scope=\\"col\\">
          Ratings
        </th>
        <th scope=\\"col\\">
          Pts
        </th>
      </tr>
    </thead>
    <tbody className=\\"criterions\\">
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={{...}} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={{...}} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
      <tr>
        <td colSpan=\\"3\\">
          <Flex justifyItems=\\"end\\" as=\\"span\\" direction=\\"row\\" alignItems=\\"center\\" inline={false} visualDebug={false} wrapItems={false}>
            <FlexItem as=\\"span\\" grow={false} shrink={false}>
              Total Points: 6.0
            </FlexItem>
          </Flex>
        </td>
      </tr>
    </tbody>
  </Table>
</div>"
`;

exports[`the Rubric component renders properly with no assessment 1`] = `
"<div className=\\"react-rubric\\">
  <Table caption=\\"Point Rubric\\" hover={false} size=\\"medium\\">
    <thead>
      <tr>
        <th scope=\\"col\\">
          Criteria
        </th>
        <th scope=\\"col\\">
          Ratings
        </th>
        <th scope=\\"col\\">
          Pts
        </th>
      </tr>
    </thead>
    <tbody className=\\"criterions\\">
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={{...}} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={{...}} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
      <tr>
        <td colSpan=\\"3\\">
          <Flex justifyItems=\\"end\\" as=\\"span\\" direction=\\"row\\" alignItems=\\"center\\" inline={false} visualDebug={false} wrapItems={false}>
            <FlexItem as=\\"span\\" grow={false} shrink={false}>
              Total Points: 11.0
            </FlexItem>
          </Flex>
        </td>
      </tr>
    </tbody>
  </Table>
</div>"
`;

exports[`the Rubric component updates the total score when an individual criterion point assessment changes 1`] = `
"<div className=\\"react-rubric\\">
  <Table caption=\\"Point Rubric\\" hover={false} size=\\"medium\\">
    <thead>
      <tr>
        <th scope=\\"col\\">
          Criteria
        </th>
        <th scope=\\"col\\">
          Ratings
        </th>
        <th scope=\\"col\\">
          Pts
        </th>
      </tr>
    </thead>
    <tbody className=\\"criterions\\">
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={[Function]} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
      <Criterion allowExtraCredit={false} assessment={{...}} criterion={{...}} customRatings={{...}} freeForm={false} isSummary={false} onAssessmentChange={[Function]} savedComments={{...}} hidePoints={false} hasPointsColumn={true} />
      <tr>
        <td colSpan=\\"3\\">
          <Flex justifyItems=\\"end\\" as=\\"span\\" direction=\\"row\\" alignItems=\\"center\\" inline={false} visualDebug={false} wrapItems={false}>
            <FlexItem as=\\"span\\" grow={false} shrink={false}>
              Total Points: 5 out of 11.0
            </FlexItem>
          </Flex>
        </td>
      </tr>
    </tbody>
  </Table>
</div>"
`;
